### ISSUE TITLE
Clink doesn't launch clink_start.cmd from custom profile folder

### ISSUE THREAD

**Issue by Rygor83** [chrisant996/clink#382]
Good morning, 

I am trying to configure clink to launch clink_start.cmd file from custom profile folder. It doesn't work. clink_start.cmd contains chcp 65001 command.

Here is clink info

![image](https://user-images.githubusercontent.com/18440565/206645370-2a1815aa-dce8-4403-aad9-9cc69d6b65ba.png)

clink.autostart value

![image](https://user-images.githubusercontent.com/18440565/206645429-98c3ecae-892a-4340-9179-9d92b6f7ba98.png)

clink custom profile folder

![image](https://user-images.githubusercontent.com/18440565/206645467-5a456235-2601-4d37-8ff0-654ce7289515.png)

clink_start.cmd in different folders. 
If clink_start.cmd is in binaries folder - everything is ok. but if it's in profile folder - nothing happens

![image](https://user-images.githubusercontent.com/18440565/206646180-11939e9e-5f56-4f40-a1ba-1856d0d239e5.png)

![image](https://user-images.githubusercontent.com/18440565/206646404-178a772d-2f11-4797-92f0-dae721644441.png)

**Comment by chrisant996** [comment]
@Rygor83 I don't think it's a Clink issue:

I've confirmed that it works as described, and as you expected.

But there's another aspect in your specific case:

Your profile is in OneDrive. Which means the clink_start.cmd file is downloaded from the Internet. So I would expect it should be getting marked as being Blocked from executing, until the Unblock button is clicked in the File Properties dialog in File Explorer.

When a file is blocked, CMD won't execute it.

I haven't set up OneDrive to test that theory, but I suspect that's what's going on here. Because certainly for other profile directories (custom or default), the script gets executed as expected.

Can you check whether the script is marked as Blocked?

**Comment by Rygor83** [comment]
Dear Chris, 

No, clink_start.cmd file is not blocked. It was not downloaded, I've created the file on my PC in onedrive folder. 

![image](https://user-images.githubusercontent.com/18440565/206842554-e4f36ab4-4433-4310-a363-fb4d7f3812ef.png)

**Comment by chrisant996** [comment]
Interesting. I've tried to reproduce the problem, but I can't -- when I try, clink_start.cmd files are run as expected from all of the possible locations (binaries dir, default profile dir, custom profile dir, or both binaries dir and profile dir).

Does it make a difference if there's a clink_start.cmd file in the binaries directory as well?

What are the contents of each of the clink_start. CMD files?

What happens if the script includes `echo reached here` as the first line? Or includes `notepad` as the first line?

(E.g. to help troubleshoot whether the script might be getting executed but output might be getting redirected somehow)

**Comment by chrisant996** [comment]
@Rygor83 I see that the profile (state) directory is a custom location, which gave me some further ideas:

- How is the profile directory pointed at the `d:\OneDrive\command_line` directory?
- What is the output from `clink autorun show`?

What if the reason `d:\onedrive\command_line\clink_start.cmd` isn't run is because the profile isn't being pointed to `d:\onedrive\command_line` until after Clink has already loaded and been initialized?

**Comment by Rygor83** [comment]
@chrisant996 , sorry for the silence.

> * How is the profile directory pointed at the `d:\OneDrive\command_line` directory?

![image](https://user-images.githubusercontent.com/18440565/208832128-4e565dae-d0ed-4f2c-9657-183f975e99f4.png)


> * What is the output from `clink autorun show`?

![image](https://user-images.githubusercontent.com/18440565/208832307-26e5efb4-2ff1-4fe4-b243-124df687846f.png)

**Comment by Rygor83** [comment]
> Does it make a difference if there's a clink_start.cmd file in the binaries directory as well?

If clink_start.cmd is only in profile directory - nothing happens, 
if clink_start.cmd is both in profile and binaries directories - then it is executed. 



> What are the contents of each of the clink_start.cmd files?

chcp 65001



> What happens if the script includes `echo reached here` as the first line? Or includes `notepad` as the first line?

If clink_start.cmd is only in profile directory - nothing happens, 
if clink_start.cmd is both in profile and binaries directories - then every command is executed: echo reached here / notepad

**Comment by chrisant996** [comment]
I found a way to reproduce the issue, and tracked down the problems (there's more than one, which is why it was hard to figure out what was happening).

## Analysis

If you press <kbd>Ctrl</kbd>-<kbd>X</kbd>,<kbd>Ctrl</kbd>-<kbd>Z</kbd> then Clink will report its internal info directly (versus launching a separate clink_x64.exe which has to backtrack to find the Clink session it's associated with and then extract info from it).

Doing that reports that the profile (state) directory is "c:\users\{current_user}\appdata\local\clink".  Which comes from the Autorun command having explicitly set it to that.

It turns out that CLINK_PROFILE is only used if the profile directory has not been set at all by any other mechanism yet.  It was documented as "takes precedence over any other mechanism", but a change in 2016 accidentally broke that.

And that's also why `clink info` reports a different profile directory than is actually being used:
- When Clink was injected, the autorun command specified a profile directory explicitly.
- When `clink info` is used, it does not specify a profile directory explicitly.  And then the bug in the precedence logic for CLINK_PROFILE isn't hit, and so it comes up with the expected profile directory ... although that isn't the _actual_ profile directory being used by the associated Clink session.  (Sheesh.)

## The change that broke it

On 2016/09/15 Martin made the following change:

**app_context.cpp**
```diff
@@  -32,8  +32,11  @@
     // The environment variable 'clink_profile' overrides all other state
     // path mechanisms.
-    if (!os::get_env("clink_profile", state_dir))
-        if (state_dir.empty())
+    if (state_dir.empty())
+        os::get_env("clink_profile", state_dir);
+
+    // Look for a state directory that's been inherited in out environment.
+    if (state_dir.empty())
-            load_from_env();
+        load_from_env();
 
     // Still no state directory set? Derive one.
     if (state_dir.empty())
```

Notice the comment didn't change and still says "clink_profile overrides all other state path mechanisms".

But the functionality changed:  it made clink_profile be used _**only**_ when a profile directory hasn't been specified by any other mechanisms.

The description of the commit is "State dir that's set externally should trump all.", but that's not how it actually behaves.

(The original logic wasn't correct either; it had a different flaw.  The good news is I see how to solve both flaws.)

## Conclusions

1. I'll fix the precedence of CLINK_PROFILE.
2. I'll confirm that fixing it solves _both_ problems.
    - Uses the CLINK_PROFILE directory as the profile directory in the Clink session.
    - Enables the `clink info` to accurately report the profile directory.

**Comment by chrisant996** [comment]
The fix is available in v1.4.6.
